Expand features to include callouts, headings, lists, and more (per #1)

elipousson opened the pull request:
Per #1 this PR includes the following new functions:

- `qto_block()` and `qto_div()`
- `qto_span()`
- `qto_dl()`, `qto_li()`, and `qto_ol()`
- `qto_heading()` and `qto_hr()`
- `qto_fig()`
- `qto_shortcode()` (and related functions)

These functions are supported by a variety of helper functions that should make it easier to continue building out the features of this package.

elipousson committed:
Add standalone type checks from rlang

elipousson committed:
Add `qto_block` and `qto_div`

Supported by `qto_attributes` helper function and new utilities in utilities.R and utilities-list.R

elipousson committed:
Add `qto_callout()`

elipousson committed:
Add `qto_heading()` and `qto_hr()`

`qto_heading()` is required for  `qto_callout()`

elipousson committed:
Add `qto_shortcode()` function

Also `qto_video()` and `qto_pagebreak()` and `qto_kbd()`

Also adds `check_src()` (used by `qto_video()`)

elipousson committed:
Add docs for `qto_shortcode()`

elipousson committed:
Add `qto_definition_list()`

elipousson committed:
Add `qto_fig()` and `qto_fig_span()` helper

elipousson committed:
Update DESCRIPTION + NAMESPACE

elipousson committed:
Style package w/ `NumSpacesForTab: 2` in `.Rproj`

elipousson committed:
Add `build_readme_qmd()` + add quarto to Suggests

Also build README

elipousson committed:
Add Eli Pousson to authors in DESCRIPTION

elipousson committed:
Add tests for `qto_block()` + `qto_div()`

elipousson committed:
Correct example for `qto_img()`

elipousson committed:
Add input check to `combine()` helper function

elipousson committed:
Add snapshot tests

elipousson committed:
Add `map_qto()` function

elipousson committed:
Fix issue w/ qto_definition_list + add .sep parameter

elipousson committed:
Mark div + span functions as internal

elipousson committed:
Drop prefixing for rlang functions

elipousson committed:
Update README

elipousson committed:
Fix issue w/ length 1 lists for qto_definition_list

elipousson committed:
Fix `qto_fig()` when caption is missing

Adds allow_null parameter to combine helper function (and related embrace + bracket + parentheses functions)

elipousson committed:
Return empty embrace "{}" by default

elipousson committed:
Fix bug w/ numeric fence values

Also

- drop sep parameter from qto_div (it didn't do anything)
- expose sep parameter for qto_block

elipousson committed:
Add ordered (qto_ol) + unordered lists (qto_li)

Also rename qto_definition_list to qto_dl

elipousson committed:
Update docs for qto_div

elipousson committed:
Add qto_span

elipousson committed:
Drop sep parameter from qto_callout

elipousson committed:
Update qto_fig_span to drop empty attribute embrace

elipousson committed:
Update snapshot tests

elipousson committed:
Update README

elipousson committed:
Remove heading parameter from qto_callout

elipousson committed:
Add span + list tests

elipousson committed:
Fix list functions

Must include a new line before list

Also expose sep parameter to switch between tight + wide list spacing + add example showing option for nested lists + expose symbol parameter for qto_ol

elipousson committed:
Restore removed fence parameter

ElianHugh commented on R/block.R:
```diff
@@ -0,0 +1,106 @@
+#' - [knitr::asis_output()]
+#'
+#' @export
+qto_block <- function(..., sep = "", collapse = "", call = caller_env()) {
```
swapping `paste0` for `paste` should make `sep` work as expected.

At the moment the following is output when trying to use sep:

```r
qto_block("A", "B", "C", sep = "e")
#> ABCe
```

ElianHugh commented on R/block.R:
```diff
@@ -0,0 +1,106 @@
+                    .attributes = NULL,
+                    .content = NULL,
+                    collapse = "",
+                    fence = ":::",
```
Is fence necessary as a parameter? I understand that it is used for changing the length of the delim, but as long as each fence has curly braces, it seems they can be nested without issue. Might be simpler to leave the fence arg off completely

```md
:::{}
:::{}
:::
:::
```

generates a nested div

elipousson replied:
I really thought there is something in the Quarto docs about requiring a sequence of four colons instead of 3 in some situation but now I can't find it. I'll keep the fence parameter internally in case we wanted to add a code block function in the future but I'll go ahead and drop it from `qto_div()`.

ElianHugh replied:
Oddly enough I recall that too. However, from the docs:

>  ... it can be helpful for _visual clarity_ to use fences of different lengths to distinguish nested divs from their parents

(emphasis mine)

I think your decision is a good compromise!

ElianHugh commented on R/basics.R:
```diff
@@ -0,0 +1,32 @@
+#' @param rule,length Horizontal rule character and length of rule.
+#' @param before,after Text to insert before and after a horizontal rule.
+#' @export
+qto_hr <- function(rule = "-",
```
Would it be worth having the "rule" parameter checked? As far as I can tell, there are only three valid values to create a rule: 

`---`
`___`
`* * *`

Then again, is it necessary to specify what the specific rule text is, given the output is identical?

elipousson replied:
I don't feel strongly about keeping it in but I can imagine a scenario where someone prefers a specific Markdown style and wants the output from these functions to be consistent with other Markdown elsewhere in a document. I went ahead and added an input check but can remove it if you think simplicity wins out over flexibility.

ElianHugh commented on R/map.R:
```diff
@@ -0,0 +1,57 @@
+#' )
+#'
+#' @export
+map_qto <- function(.x,
```
Just to double check, is the output expected to print as markdown? At present it outputs a list object, which won't print `asis` on its own. Ideally I think it's a good idea to wrap objects output by `quartools` with the `qto_block` function to make it easier on users. I could see it being handy, however, to iterate over lists of markdown content, so I thought I would check what the intent was!

This also makes me think it might be useful to have a list-type quarto block class...

elipousson replied:
I kept the list output for type stability but I'll add an example showing how to splice the list output into `qto_block()` or collapse a Quarto block list output. I was thinking of adding a print method that automatically collapsed lists into Quarto blocks but that may be too opinionated/risky of a change.

ElianHugh replied:
Exactly what I was thinking re: custom print method. Perhaps something that maybe should be pushed to a later PR/issue for discussion.

All good with the use case here!

ElianHugh commented on R/figures.R:
```diff
@@ -0,0 +1,79 @@
+      ref_prefix <- "#fig-"
+    }
+
+    reference <- paste0(prefix, reference)
```
Should this be `ref_prefix` not `prefix`?

elipousson replied:
Good catch!

ElianHugh commented on R/attributes.R:
```diff
@@ -0,0 +1,116 @@
+      if (is.logical(x)) {
+        return(tolower(x))
+      }
+      # FIXME: Percent values are enclosed in quotes, e.g. 60%, but should not
```
A super basic check could be:

```r
if (endsWith(x, "%")) {
  mark <- ""
}
```

...Or a regex if we want to be more precise in targetting actual percentages, e.g. 

```r
if (grepl(r"{\b[0-9]+\.?[0-9]*%\b}", x)) {
	mark <- ""
}
```

ElianHugh commented:
Thank you for the PR, this looks great! I've had a short pass over the build with a few comments, but happy to have a discussion on anything. 

Before committing I'd like to do a final pass to drop some old functions, fix some lintr issues, and fix some RCMD check issues

Edit: whoops. I forgot to actually hit the submit review button :)

elipousson committed:
Remove fence parameter from `qto_div()`

elipousson committed:
Add check for rule input

elipousson committed:
Fix handling of reference prefix

elipousson committed:
Allow unnamed attributes

Also fix handling of percent attributes

elipousson committed:
Clean up examples for qto_block + qto_div

elipousson committed:
Remove fence arg from `qto_callout()`

elipousson committed:
Add docs for combine

Also tweak build_readme_qmd to better handle default path

elipousson committed:
Add docs for combine

elipousson committed:
Fill in missing attribute definitions

elipousson committed:
Allow formula inputs for .f

Also expose sep and collapse arguments for qto_block

elipousson committed:
Fill in missing parameter definitions

elipousson committed:
Add examples for `qto_dl()`

Fix links in `qto_ol()` + `qto_li()` docs

elipousson commented:
I think I resolved all of the issues you flagged in your review and filled in a lot of missing documentation so cleaning up the check messages should be a good bit faster. I left the feedback on `qto_hr()` and `map_qto()` unresolved just in case there you wanted further changes to the former or to open an issue for the latter before we call it done.

Hopefully, I didn't introduce any new issues with the documentation clean-up! The only item I haven't done is add more tests but I figured you may either have suggestions on how you want that done or you may want to add them yourself as part of that "final pass". 

One suggestion for `mdapply()`: I wonder if reworking it as a `pmap_qto()` function might be a good approach – should work much the same and provide a parallel to `map_qto()`.

ElianHugh committed:
Linter pass

- Add standalones to lintr ignore
- Use explicit integers
- Match lintr indentation
- Implicit return

ElianHugh committed:
Mark mdapply and as_markdown as internal

ElianHugh committed:
Fix attr typo

- Fix description typo of class and id using the opposite identifier

ElianHugh commented:
Hi Eli, thanks again for submitting the PR. I've just pushed some minor changes for style/linter appeasement, and fixed a few typos. I've also marked `mdapply` and `as_markdown` as internal, and will shortly follow up with another PR implementing:

1. `pmap_qto`
2. removal of `div`, `span`, `mdapply`, and `as_markdown` functions
3. change the with_* functions to use `qto_div` instead of the superceded `div` function

If you're happy with how things are with the PR, I'll go ahead and squash merge it!

ElianHugh committed:
Update docs

elipousson commented:
Looks good to me. Thanks!
